mavros_extras/src/plugins/odom.cpp

```diff
@@ -66,88 +67,92 @@ class OdometryPlugin : public plugin::PluginBase {
+			tf_parent2ned = tf2::transformToEigen(m_uas->tf2_buffer.lookupTransform(
+				odom->header.frame_id, "local_origin_ned", ros::Time(0)));
+		} catch (tf2::TransformException &ex) {
+			ROS_WARN("%s",ex.what());
+			ros::Duration(1.0).sleep();
+			return;
+		}
+
+		// unpack data into Eigen
+		Eigen::Vector3d pos_parent(
+			odom->pose.pose.position.x,
+			odom->pose.pose.position.y,
+			odom->pose.pose.position.z);
+
+		Eigen::Vector3d lin_vel_child(
+			odom->twist.twist.linear.x,
+			odom->twist.twist.linear.y,
+			odom->twist.twist.linear.z);
+
+		Eigen::Vector3d ang_vel_child(
+			odom->twist.twist.angular.x,
+			odom->twist.twist.angular.y,
+			odom->twist.twist.angular.z);
+
+		Eigen::Quaterniond q_parent2child(
+			odom->pose.pose.orientation.w,
+			odom->pose.pose.orientation.x,
+			odom->pose.pose.orientation.y,
+			odom->pose.pose.orientation.z);
+
+		Eigen::Vector3d lin_vel_NED;
+		Eigen::Vector3d ang_vel_ned;
+		Eigen::Vector3d pos_NED;
+		Eigen::Quaterniond q_parent2ned;
+
+		Eigen::Matrix<double, 6, 6> pose_cov_parent;
+		Eigen::Matrix<double, 9, 9> pose_cov_ned;
+		Eigen::Matrix<double, 6, 6> twist_cov_child;
+		Eigen::Matrix<double, 3, 3> att_cov_ned;
+
+		for (int i=0; i<6; i++) for (int j=0; j<6; j++) {
+			pose_cov_parent(i, j) = odom->pose.covariance[i*6 + j];
+			twist_cov_child(i, j) = odom->pose.covariance[i*6 + j];
+		}
+
+		// apply transforms
+		lin_vel_NED = tf_child2ned.linear()*lin_vel_child;
+		ang_vel_ned = tf_child2ned.linear()*ang_vel_child;
+		pos_NED = tf_parent2ned.linear()*pos_parent;
+		q_parent2ned = tf_parent2ned.linear()*q_parent2child;
+
+		// constrauct 9x6 matrix to rotate pose from parent to NED
+		// note it is 9x6 since the lpos matrix has accel states
+		// while ROS odometry message doesn't
+		Eigen::Matrix<double, 9, 6> RPose;
+		RPose.block<3, 3>(0, 0) = tf_parent2ned.linear();
+		RPose.block<3, 3>(3, 3) = tf_parent2ned.linear();
 
 		// get timestamp
 		uint64_t stamp_usec = odom->header.stamp.toNSec() / 1e3;
-		const auto zerov3f = Eigen::Vector3f::Zero();
 
 		/* -*- LOCAL_POSITION_NED_COV parser -*- */
 		mavlink::common::msg::LOCAL_POSITION_NED_COV lpos {};
 
 		lpos.time_usec = stamp_usec;
+		pose_cov_ned = RPose*pose_cov_parent*RPose.transpose();
 		lpos.estimator_type = utils::enum_value(estimator_type);
-		ftf::covariance9d_urt_to_mavlink(cov_full_tf, lpos.covariance);
-
-		// [[[cog:
-		// for a, b in (('', 'pos_ned'), ('v', 'lin_vel_ned'), ('a', 'zerov3f')):
-		//     for f in 'xyz':
-		//         cog.outl("lpos.{a}{f} = {b}.{f}();".format(**locals()))
-		// ]]]
-		lpos.x = pos_ned.x();
-		lpos.y = pos_ned.y();
-		lpos.z = pos_ned.z();
-		lpos.vx = lin_vel_ned.x();
-		lpos.vy = lin_vel_ned.y();
-		lpos.vz = lin_vel_ned.z();
-		lpos.ax = zerov3f.x();
-		lpos.ay = zerov3f.y();
-		lpos.az = zerov3f.z();
-		// [[[end]]] (checksum: 4620cfc77f3e5d768d7c23da5f56ddc6)
+		for (int i=0; i<6; i++) for (int j=0; j<6; j++) {
+			lpos.covariance[i*6 + j] = pose_cov_ned(i, j);
+		}
```

Comment:
According to [mavlink docs][1] covariance is 9x9 not 6x6, and only upper triangular part.

[1]: http://mavlink.org/messages/common#LOCAL_POSITION_NED_COV